### ISSUE TITLE
Stop using OpenSSL-license-licensed code.

### ISSUE THREAD

**Issue by tisonkun** [briansmith/ring#1827]
Currently, the [LICENSE](https://github.com/briansmith/ring/blob/6c29bf61cd0e9d478deeec11e922e8513d5d0f3c/LICENSE) file is large and complicated, while can be vaguely regarded as a combination of MIT, ISC, and OpenSSL’s licenses.

Although [the latest OpenSSL license is Apache License 2.0](https://www.openssl.org/source/apache-license-2.0.txt), in this repo, we use a few of sentences refer to the BSD-style licenses with advertising clauses:

https://github.com/briansmith/ring/blob/6c29bf61cd0e9d478deeec11e922e8513d5d0f3c/LICENSE#L66-L83

https://github.com/briansmith/ring/blob/6c29bf61cd0e9d478deeec11e922e8513d5d0f3c/LICENSE#L137-L145

It causes concerns in downstream for using this software in a mindset like so-called "permissive OSS license" or "weak copyleft license": https://lists.apache.org/thread/ptwdv18z4wd9r11nmdwj7wgzwvm3b8l2

@briansmith do you have more background of the license content, or how generally a downstream user use it?

The background of this questions is from https://www.apache.org/legal/resolved.html#category-x that "BSD-4-Clause/BSD-4-Clause (University of California-Specific)" can introduce burden for users to convey this software - they're, be required, to include extra acknowledgement for certern actions. And while if we can either:

1. use an Apache License 2.0 base, or;
2. be clear that what part of this crate is not covered by these licenses.

**Comment by briansmith** [comment]
I believe it is already clear which parts of the crate are covered by which license as each source file has a license header at the top. The code in question comes from BoringSSL and they didn't do the relicensing. Unfortunately they (BoringSSL) put some important bits in OpenSSL-license-licensed header files. Our goal is to replace all the OpenSSL-license-licensed code as we move completely away from C to Rust. Once that is done there won't be any issue regarding this anymore.

See also #902 which tracks making the licensing use SPDX.

**Comment by tisonkun** [comment]
@briansmith Thanks for your reply. I have one more question:

> it is already clear which parts of the crate are covered by which license

For a takeaway, are those files OpenSSL-license-licensed included in any downstream software that use ring? Or it's optional if certain feature unused.

I did a check and it seems the fileset is:

* [ ] crypto/constant_time_test.c
* [ ] crypto/cpu_intel.c
* [ ] crypto/internal.h
* [ ] crypto/mem.c
* [ ] crypto/fipsmodule/bn/internal.h
* [ ] crypto/fipsmodule/bn/montgomery.c
* [ ] include/ring-core/aes.h
* [ ] include/ring-core/arm_arch.h
* [ ] include/ring-core/base.h
* [ ] include/ring-core/mem.h
* [ ] include/ring-core/type_check.h

You can check if there is other missing, and perhaps we can divide and conquer them :D

**Comment by briansmith** [comment]
> > include/ring-core/aes.h
> We need to vendor the rust-crypto aes/soft code at https://github.com/RustCrypto/block-ciphers/tree/master/aes/src/soft into src/third_party/rust_crypto/aes/soft, replacing aes_nohw.c. Then we can remove aes_nohw.c and aes.h.

This is now tracked as #1886 with more details.

**Comment by nyabinary** [comment]
Yeah, it makes using the crate practically impossible in GPL licensed stuff:
https://www.gnu.org/licenses/license-list.en.html#OriginalBSD
Nice to see progress being made on this tho :3

**Comment by SDAChess** [comment]
Thank you for your hard work on this :) greatly appreciated!

**Comment by SDAChess** [comment]
If there is anything I can help with regarding this effort; let me know I would love to contribute.

**Comment by briansmith** [comment]
I would love it if people could review PR #2346 which replaces one of the most critical OpenSSL-license-licensed source files with brand new Rust code with the standard *ring* license.

**Comment by briansmith** [comment]
Besides recent work that has gone into reducing the amount of C code used, upstream BoringSSL recently relicensed everything to the OpenSSL Apache 2.0 license. So the worst case scenerio is that whatever doesn't make sense to replace will get converted from OpenSSL license to Apache license in an upcoming merge of BoringSSL into *ring*.

**Comment by briansmith** [comment]
With PR #2397, the vast majority of the changes for this have been done. The LICENSE file hasn't changed yet but that will happen soon.

**Comment by briansmith** [comment]
I merged PR #2402, which I think resolves this issue. PLMK if there are things I overlooked.

**Comment by briansmith** [comment]
PR #2406 further clarifies licensing.

**Comment by tisonkun** [comment]
Great! Thank you.
